### PR TITLE
Transform mixed type/value imports to type-only imports

### PR DESCRIPTION
**Pull request by dwb**
```js
import { type A, B } from "./a"
```
is invalid TypeScript, so we transform it into two statements, one
being a [type-only import](https://www.typescriptlang.org/docs/handbook/modules.html#importing-types).

Also updated `trackComments` to handle multiple nodes, which you
might have passed to `replaceWithMultiple`.

### CODE REVIEW

**Comment by k4b7** [comment]
@dwb thanks for the PR!  You'll need to sign our [Contributor License Agreement](http://khanacademy.org/r/cla) before I can accept this contribution.  Please comment here once you've done so.

**Comment by dwb** [comment]
@kevinbarabash Hi Kevin, thanks – we've signed the corporate CLA as Monzo Bank with me (and soon one other engineer) listed 🙏

**Comment by k4b7** [comment]
@dwb thanks!  I'll review the PR this evening. 🙂

File: src/transform.ts [hunk 1]
```diff
@@ -575,31 +576,89 @@ export const transform = {
...
-        path.replaceWith(replacementNode);
-      } else {
-        trackComments(path.node, state);
+        if (typeSpecs.length > 0) {
+          const typeNode = t.importDeclaration(
+            typeSpecs.map((s) =>
+              t.importSpecifier(
+                returning(t.identifier(s.local.name), (n) => {
+                  n.loc = s.local.loc;
+                }),
```
Comment (k4b7) [inline src/transform.ts:618]: style nit: this is clever, but it's more general than it needs to be which ends up making it hard to follow.  I think a more focused helper would be cleaner and easier to grok:
```
withLocFrom(t.identifier(s.local.name), s.local);
```

File: src/transform.ts [hunk 2]
```diff
@@ -575,31 +576,89 @@ export const transform = {
...
+      }
+
+      trackComments(replacementNode ?? path.node, state);
+
+      if (replacementNode != null) {
+        if (Array.isArray(replacementNode)) {
+          path.replaceWithMultiple(replacementNode);
+        } else {
+          path.replaceWith(replacementNode);
+        }
```
Comment (k4b7) [inline src/transform.ts:661]: style nit: It feels like drying up the code made things more complicated.  Could you move the `trackComments()` and `replaceWith()`/`replaceWithMultiple()` cals back into the if/else-if/else blocks?

File: src/util.ts [hunk 3]
```diff
@@ -20,28 +20,53 @@
...
 export const trackComments = (node, state) => {
-  if (node.leadingComments) {
-    for (const comment of node.leadingComments) {
+  let leadingNode = node,
+    trailingNode = node;
+
+  if (Array.isArray(node)) {
+    leadingNode = node[0];
+    trailingNode = node[node.length - 1];
+  }
```
Comment (k4b7) [inline src/util.ts:29]: style nit: this could be simplified to:
```
const [leadingNode, trailingNode] = Array.isArray(node)
  ? [node[0], node[node.length - 1]]
  : [node, node];
```

**Review by k4b7 — APPROVED** [review]
Thanks for the PR.  I appreciate you taking the effort to handle comments properly.  I left a few style nits in the comments.  Let me know what you think about them.

File: src/transform.ts (hunk 1 above)
Comment (dwb) [reply src/transform.ts:618]: Up to you – personally I prefer these sort of general but also straight-forward combinators, but it's a matter of taste really.

With your suggestion I would expect the order of those arguments to be the other way around, though?

File: src/util.ts (hunk 3 above)
Comment (dwb) [reply src/util.ts:29]: Monzo generally dislikes the ternary operator so this was much more that influence than anything! I'm ambivalent so happy to change it.

File: src/transform.ts (hunk 1 above)
Comment (k4b7) [reply src/transform.ts:618]: That's a good point about the arg order.

File: src/util.ts (hunk 3 above)
Comment (k4b7) [reply src/util.ts:29]: At KA the preference is to use `const` as much as possible which leads to more ternary operators in the code.

**Comment by k4b7** [comment]
@dwb I've published 0.5.2 with your change in it.  Thanks again for the PR. 🙂

**Comment by dwb** [comment]
Thanks! To be clear I was happy to make those changes but yes, I don't
think either of us thought they were super important either way :)

On April 27, 2021, GitHub Notifications ***@***.***>
wrote:
> @dwb <https://github.com/dwb> I've published 0.5.2 with your change in
> it. Thanks again for the PR. 🙂
>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/Khan/flow-to-ts/pull/273#issuecomment-827837857>, or
unsubscribe <https://github.com/notifications/unsubscribe-
auth/AABFBK5EZIOSST3NWIX34ZTTK4BTRANCNFSM423UBVOA>.
